File: classes/PTVFormatter.py
import datetime
import json
import grequests

import discord
from ptv.client import PTVClient

class PTVFormatter(PTVClient):
	def __init__(self, dev_id, api_key, not_secure = False):
		super().__init__(dev_id, api_key, not_secure=not_secure)

	def _splicegen(self, maxchars, stringlist):
		runningcount = 0
		tmpslice = []
		for item in stringlist:
			runningcount += len(item)
			if runningcount <= int(maxchars):
				tmpslice.append(item)
			else:
				tmpslice = [item]
				runningcount = len(item)
		yield(tmpslice)

	def _groupDisruptions(self, disruptions):
		dict1 = {"Part suspended" : [],
				"Major Delays" : [], 
				"Minor Delays" : [], 
				"Planned Works" : [], 
				"Planned Closure" : [], 
				"Service Information" : [],
				"Other" : []}
		for disruption in disruptions:
			if disruption['disruption_type'] in dict1:
				dict1[disruption['disruption_type']].append(f"[{disruption['title']}]({disruption['url']})")
			else:
				dict1["Other"].append(f"[{disruption['title']}]({disruption['url']})")
		return dict1

	def setDisruptionColour(self, disruptionsmsg, route_id, bot, disruptions):
		ServiceStatusColours = {
				"Good Service" : 0x97d700, 
				"Service Information" : 0xd3ecf4, 
				"Major Delays" : 0xef4135, 
				"Minor Delays" : 0xe87800, 
				"Planned Works" : 0xffd500, 
				"Planned Closure" : 0xffd500, 
				"Part suspended" : 0x1f1f1f,
				"Other" : 0x97d700}
		if disruptions["Part suspended"]:
			disruptionsmsg.colour = 0x1f1f1f
			return
		try:
			disruptionsmsg.colour = ServiceStatusColours[bot.routeservicestatus["routes"][int(route_id) - 1 if int(route_id) < 10 else int(route_id) - 2]["route_service_status"]["description"]]
		except KeyError:
			disruptionsmsg.colour = 0x97d700

	def disruptions_to_embed(self, disruptionsmsg, disruptions, route_id, bot):
		"""
		Changes Disruptions from PTV Api to an Embed object

		Parameters
		----------
		Disruptionsmsg : embedobject
			Embed object of the disruptions msg
		disruptions : dict
			dict of disruptions from PTV Api

		Returns
		-------
		Disruptionsmsg : embedobject
			Embed object of completed disruptions msg
		"""
		disruptions = self._groupDisruptions(disruptions)
		if bot.routeservicestatus["routes"][int(route_id) - 1 if int(route_id) < 10 else int(route_id) - 2]["route_service_status"]["description"] == "Good Service" and not disruptions["Part suspended"]:
			disruptionsmsg.add_field(name="Good Service", value='Trains are currently running on time to five minutes.')
		self.setDisruptionColour(disruptionsmsg, route_id, bot, disruptions)
		for disruptiontype, disruption in disruptions.items():
			if disruption:
				splicedisruption = self._splicegen(1024, disruption)
				for i in splicedisruption:
					# print(i)
					disruptionsmsg.add_field(name=disruptiontype, value='\n'.join(i))
		return disruptionsmsg

	def UpdateDisruptions(self, bot):
		disruptions = self.get_disruptions()["disruptions"]["metro_train"]
		bot.routeservicestatus = self.get_routes(route_types=0)
		newdisruptions = {"1":[], "2":[],"3":[],"4":[],"5":[],"6":[],"7":[],"8":[],"9":[],"10":[],"11":[],"12":[],"13":[],"14":[],"15":[],"16":[],"17":[]}
		notuptodatedisruptions = []
		for disruption in disruptions:
			for route in disruption["routes"]:
				try:
					newdisruptions[str(route["route_id"])].append(disruption)
				except KeyError:
					pass
		with open('./InfoFiles/disruptions.json','r') as f:
			previousdisruptions = json.load(f)
		with open('./InfoFiles/disruptions.json','w') as f:
			f.write(json.dumps(newdisruptions))
		for key, value in newdisruptions.items():
			if previousdisruptions[key] != value:
				notuptodatedisruptions.append(key)
		return newdisruptions, notuptodatedisruptions

	def getDepartureInfos(self, station):
		data = {}
		# with open('./InfoFiles/directions.json', 'r') as f:
		# 	directionss = json.load(f)
		# with open('./InfoFiles/stations.json', 'r') as f:
		# 	stations = json.load(f)
		# station = Stations[station.upper()].value
		# station = station.title()
		# try:
		#	 station = stations[station.replace(' Station', '')]
		# except KeyError:
		#	 return('Error, station not found')
		data['Station'] = station
		# print(Directions)
		Directions = self._callApi(f"/v3/departures/route_type/{station['route_type']}/stop/{station['stop_id']}", params={"expand":["all"]})
		data['Directions'] = Directions['directions']
		# for route in station['routes']:
		#	 for direction in directionss[str(route)]:
		#		 if not direction in data['Directions']:
		#			 data['Directions'].append(direction)
		stopid = station['stop_id']
		urls = []
		for direction_id, direction in data['Directions'].items():
			urls.append(self._getUrl(f"/v3/departures/route_type/{station['route_type']}/stop/{stopid}", params={"direction_id":direction_id, "max_results":3, "expand":["run"]}))
			# NextTrain = PTV.get_departure_from_stop(RouteType.TRAIN, stopid, direction_id=direction['direction_id'], max_results=3).get('departures')[0:3]
		urls.append(self._getUrl(f"/v3/disruptions/stop/{stopid}", params={"disruption_status":'current'}))
		rs = (grequests.get(u) for u in urls)
		rs = grequests.map(rs)
		data['Departures'] = []
		for i in rs[:-1]:
			i = i.json()
			i['departures'] = i['departures'][0:3]
			i['runs'] = i['runs']
			data['Departures'].append(i)
		data['Disruptions'] = (rs[-1].json())
		print(data)
		return data
		

File: functions/PTVFormatter.js
const { MessageActionRow, MessageSelectMenu, MessageEmbed, DiscordAPIError, MessageAttachment } = require('discord.js');
const ptvApi = require('../functions/PTVApi');
const tfnswApi = require('../functions/TfNSWApi');

const fs = require('fs');

const avatar_url = 'https://cdn.discordapp.com/avatars/503096810961764364/f89dad593aa8635ccddd3d364ad9c46a.png'

class ptvFormatter{
    ptvEmojiRoute = [
        {name: "Metro", id: "771920140207259659"},
        {name: "Tram", id: "771921271998382140"},
        {name: "Bus", id: "771921102335246346"},
        {name: "VLine", id: "771920567959683102"},
        {name: "Bus", id: "771921102335246346"}
    ]

    tfnswRoutes = {
        1: {name: "TfNSW_T", id: "988139276275449967", color: "#F47A0B"},
        2: {name: "TfNSW_M", id: "988144694892568607", color: "#00969f"},
        4: {name: "TfNSW_L", id: "988139544245329971", color: "#EE1F35"},
        5: {name: "TfNSW_B", id: "988139592291057684", color: "#00A8E6"},
        9: {name: "TfNSW_F", id: "988139631075799040", color: "#50BF33"}
    }

    PTVColours = [
        "#0072ce",
        "#78be20",
        "#ff8200",
        "#8f1a95",
        "#ff8200"
    ]
    
    ServiceStatusColours = {
        "Good Service": 0x97d700, 
        "Service Information": 0xd3ecf4, 
        "Major Delays": 0xef4135, 
        "Minor Delays": 0xe87800, 
        "Planned Works": 0xffd500, 
        "Planned Closure": 0xffd500, 
        "Part suspended": 0x1f1f1f,
        "Other": 0x97d700
    }
        
    // Route Type to English Translation with singular and plural versions
    PTVRouteTypeTranslate = [
        ["Metro","Metro Trains"],
        ["Tram","Trams"],
        ["Bus","Busses"],
        ["VLine", "VLine Trains"],
        ["Bus", "Busses"]
    ]

    TfNSWRouteTypeTranslate = {
        1: ["Train","Trains"],
        2: ["Metro","Metro Trains"],
        4: ["Light Rail Tram","Light Rail Trams"],
        5: ["Bus","Busses"],
        7: ["Coach","Coaches"],
        9: ["Ferry","Ferries"],
        11: ["School Bus","School Busses"]
    }

    constructor(ptvDevId, ptvApiKey, tfnswApiKey) {
        this.ptvClient = new ptvApi(ptvDevId, ptvApiKey);
        this.tfnswClient = new tfnswApi(tfnswApiKey)
    }

    async searchToMenu(search_term, route_types) {
        
        const ptv_search_results = await this.ptvClient.searchForStop(search_term, route_types)
        const tfnsw_search_results = await this.tfnswClient.searchForStop(search_term)

        var stops = []

        for (var i = 0; i < Math.min(ptv_search_results['stops'].length, 13); i++) {
            stops.push({
                label: ptv_search_results['stops'][i]['stop_name'],
                description: ptv_search_results['stops'][i]['stop_suburb'],
                value: "VIC" + String(ptv_search_results['stops'][i]['route_type']) + String(ptv_search_results['stops'][i]['stop_id']), // Concatonates route type with stop id to prevent error occuring with two stops of the same id (different route types)
                emoji: this.ptvEmojiRoute[ptv_search_results['stops'][i]['route_type']]
            })
        }

        for (var i = 0; i < Math.min(tfnsw_search_results['locations'].length, 12); i++) {
            for (var mode of tfnsw_search_results['locations'][i]['modes']) {
                if (mode == 11) {continue;}
                stops.push({
                    label: tfnsw_search_results['locations'][i]['disassembledName'],
                    description: tfnsw_search_results['locations'][i]['parent']['name'],
                    value: "NSW" + String(mode) + tfnsw_search_results['locations'][i]['id'],
                    emoji: this.tfnswRoutes[mode]
                })
            }
        }

        const menu = new MessageActionRow()
            .addComponents(
                new MessageSelectMenu()
                    .setCustomId('stop_select')
                    .setPlaceholder('Select Your Station')
                    .addOptions(stops),
            );

        return menu
    }

    async ptvStopToDeparturesEmbed(stop_id, route_type) {
        
        const departure_results = await this.ptvClient.getDepartures(stop_id, route_type);

        // Sort Departures into dictionary with key direction_id and value being a list of departures with that direction id

        var departures = {}

        for (var departure of departure_results['departures']) {
            if (!departures[departure['direction_id']]) {
                departures[departure['direction_id']] = [];
            }
            departures[departure['direction_id']].push(departure)
        }

        // Convert departures into an Embed

        // const thumbnail = new MessageAttachment('./src/Icons/' + this.PTVRouteTypeTranslate[route_type][0] + '.png')

        var departuresEmbed = new MessageEmbed()
            .setTitle('Next ' + this.PTVRouteTypeTranslate[route_type][0])
            .setColor(this.PTVColours[route_type])
            .setDescription(
`Next ${this.PTVRouteTypeTranslate[route_type][1]} departing from ${departure_results['stops'][stop_id]['stop_name']}.
Train Types are sourced from ptv however seem to be quite inaccurate in some cases. They may or may not be accurate.`)
            .setAuthor('VPT Bot', avatar_url)
            // .attachFiles(thumbnail)
            .setThumbnail('https://raw.githubusercontent.com/lucky962/PTBot/main/src/Icons/' + this.PTVRouteTypeTranslate[route_type][0] + '.png')
            .setFooter('Source: Licensed from Public Transport Victoria under a Creative Commons Attribution 4.0 International Licence.', avatar_url)
        
        for (var direction in departures) {
            var direction_text = '';
            var noofdepartures = 0;
            for (var departure of departures[direction]) {
                var flags = '';
                if (departure['flags'].includes('S_WCA')) {
                    flags += ("<:WCA:780582086398181386> ")
                }
                if (departure['flags'].includes('S_VTR')) {
                    flags += ("VLine Train")
                }
                if (departure['flags'].includes('S_VCH')) {
                    flags += ("VLine Coach")
                }
                direction_text = 
`${direction_text}

Plat: ${departure['platform_number']}. ${(departure_results['runs'][departure['run_ref']]['vehicle_descriptor']) ? departure_results['runs'][departure['run_ref']]['vehicle_descriptor']['description'] : ""}
Scheduled: <t:${(new Date(departure['scheduled_departure_utc'])).getTime() / 1000}:${((new Date(departure['scheduled_departure_utc'])).toDateString() == (new Date()).toDateString()) ? 't' : 'f'}>
ETA: ${(departure['estimated_departure_utc']) ? ('<t:' + (new Date(departure['estimated_departure_utc'])).getTime() / 1000 + ':R>') : 'None'}
Flags:${flags}`
                noofdepartures += 1;
                if (noofdepartures > 2) {
                    break;
                }
            }
            departuresEmbed.addField(departure_results['directions'][direction]['direction_name'], direction_text, true)
        }
        
        // Convert Disruptions into an Embed

        var disruptions = '';
        
        var disruptionsEmbed = new MessageEmbed()
            .setTitle('Potential Disruptions')
            .setColor(this.PTVColours[route_type])
            .setDescription('Potential Disruptions that might affect ' + departure_results['stops'][stop_id]['stop_name'])
            .setAuthor('VPT Bot', avatar_url)
            .setThumbnail('https://raw.githubusercontent.com/lucky962/PTBot/main/src/Icons/' + this.PTVRouteTypeTranslate[route_type][0] + '.png')
            .setFooter('Source: Licensed from Public Transport Victoria under a Creative Commons Attribution 4.0 International Licence.', avatar_url)

        for (var disruption in departure_results['disruptions']) {
            if (disruptions.length + `[${departure_results['disruptions'][disruption]['title']}](${departure_results['disruptions'][disruption]['url']})\n`.length <= 1024) {
                disruptions += `[${departure_results['disruptions'][disruption]['title']}](${departure_results['disruptions'][disruption]['url']})\n`
            } else {
                disruptionsEmbed.addField('Potential Disruptions', disruptions);
                disruptions = '';
            }
        }

        if (disruptionsEmbed.fields.length == 0) {

            const departuresMessage = {content: 'Departures:', embeds: [departuresEmbed]}
    
            return departuresMessage
        }

        if (disruptions != '') {
            disruptionsEmbed.addField('Potential Disruptions', disruptions)
        }

        if (departuresEmbed.length + disruptionsEmbed.length > 6000) {
            var disruptions = '';
            
            var disruptionsEmbed = new MessageEmbed()
                .setTitle('Potential Disruptions')
                .setColor(this.PTVColours[route_type])
                .setDescription('Potential Disruptions that might affect ' + departure_results['stops'][stop_id]['stop_name'])
                .setAuthor('VPT Bot', avatar_url)
                .setThumbnail('https://raw.githubusercontent.com/lucky962/PTBot/main/src/Icons/' + this.PTVRouteTypeTranslate[route_type][0] + '.png')
                .setFooter('Source: Licensed from Public Transport Victoria under a Creative Commons Attribution 4.0 International Licence.', avatar_url)
    
            for (var disruption in departure_results['disruptions']) {
                if (disruptions.length + `${departure_results['disruptions'][disruption]['title']}\n`.length <= 1024) {
                    disruptions += `${departure_results['disruptions'][disruption]['title']}\n`
                } else {
                    disruptionsEmbed.addField('Potential Disruptions', disruptions);
                    disruptions = '';
                }
            }

        }

        if (disruptions != '') {
            disruptionsEmbed.addField('Potential Disruptions', disruptions)
        }

        const departuresMessage = {content: 'Departures:', embeds: [departuresEmbed, disruptionsEmbed]}

        return departuresMessage
    }

    async nswStopToDeparturesEmbed(stop_id, route_type) {
        
        const departure_results = await this.tfnswClient.getDepartures(stop_id, route_type);

        var noofdepartures = {}
        var departures = {}

        for (var departure of departure_results['stopEvents']) {
            if (noofdepartures[departure['transportation']['destination']['name']] >= 2) {
                continue;
            }
            if (departure['transportation']['destination']['name'] in noofdepartures) {
                noofdepartures[departure['transportation']['destination']['name']] = noofdepartures[departure['transportation']['destination']['name']] + 1
            } else {
                noofdepartures[departure['transportation']['destination']['name']] = 0
            }
            departures[departure['transportation']['destination']['name']] = 
`${(departures[departure['transportation']['destination']['name']]) ? departures[departure['transportation']['destination']['name']] : ''}

${(departure['location']['name'].split(', '))[departure['location']['name'].split(', ').length - 1]}
Scheduled: <t:${(new Date(departure['departureTimePlanned'])).getTime() / 1000}:${((new Date(departure['departureTimePlanned'])).toDateString() == (new Date()).toDateString()) ? 't' : 'f'}>
ETA: ${(departure['departureTimeEstimated']) ? ('<t:' + (new Date(departure['departureTimeEstimated'])).getTime() / 1000 + ':R>') : 'None'}
Flags:${(departure['properties']['WheelchairAccess'])?'<:WCA:780582086398181386>':''}`
        }

        var departuresEmbed = new MessageEmbed()
            .setTitle('Next ' + this.TfNSWRouteTypeTranslate[route_type][1])
            .setColor(this.tfnswRoutes[route_type]['color'])
            .setDescription(
`Next ${this.TfNSWRouteTypeTranslate[route_type][1]} departing from ${departure_results['locations'][0]['name']}.`)
            .setAuthor('VPT Bot', avatar_url)
            // .attachFiles(thumbnail)
            .setThumbnail('https://raw.githubusercontent.com/lucky962/PTBot/main/src/Icons/' + this.tfnswRoutes[route_type]['name'] + '.png')
            .setFooter('Source: TfNSW API', avatar_url)

        for (var destination in departures) {
            departuresEmbed.addField(destination, departures[destination], true)
        }

        const departuresMessage = {content: 'Departures:', embeds: [departuresEmbed]}

        return departuresMessage
    }

    async getDisruptions() {
        
        const disruptions_results = await this.ptvClient.getDisruptions();

        var disruptions = {1:{}, 2:{},3:{},4:{},5:{},6:{},7:{},8:{},9:{},10:{},11:{},12:{},13:{},14:{},15:{},16:{},17:{}};

        for (var disruption in disruptions) {
            disruptions[disruption] = {"Part suspended" : [],
            "Major Delays" : [], 
            "Minor Delays" : [], 
            "Planned Works" : [], 
            "Planned Closure" : [], 
            "Service Information" : []}
        }

        for (var disruption of disruptions_results['disruptions']['metro_train']) {
            for (var route of disruption['routes']) {
                if (!disruptions[route['route_id']]) {
                    disruptions[route['route_id']] = {"Part suspended" : [],
                    "Major Delays" : [], 
                    "Minor Delays" : [], 
                    "Planned Works" : [], 
                    "Planned Closure" : [], 
                    "Service Information" : []}
                }

                if (!(disruption['disruption_type'] in disruptions[route['route_id']])) {
                    disruptions[route['route_id']][disruption['disruption_type']] = [];
                }
                disruptions[route['route_id']][disruption['disruption_type']].push(disruption);
                // disruptions[route['route_id']]['route_name'] = route['route_name'];
            }
        }

        return disruptions
    }

    async disruptionsToEmbed(disruptions) {

        const routes = await this.ptvClient.getRoutes();

        var disruptionsEmbeds = {};

        for (var route of routes['routes']) {
            if ([1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 17].includes(route['route_id'])) {
                var disruptionEmbed = new MessageEmbed()
                .setTitle(route['route_name'])
                .setFooter('Last Disruption Update', avatar_url)

                if (route['route_service_status']['description'] in this.ServiceStatusColours) {
                    disruptionEmbed.setColor(this.ServiceStatusColours[route['route_service_status']['description']])
                }

                if (route['route_service_status']['description'] == 'Good Service') {
                    disruptionEmbed.addField('Good Service', 'Trains are currently running on time to five minutes.')
                }

                for (var disruption_type in disruptions[route['route_id']]) {
                    if (disruptions[route['route_id']][disruption_type].length !== 0) {
                        var disruptionsTxt = '';
                        for (var disruption in disruptions[route['route_id']][disruption_type]) {
                            if (disruptionsTxt.length + `[${disruptions[route['route_id']][disruption_type][disruption]['title']}](${disruptions[route['route_id']][disruption_type][disruption]['url']})\n`.length <= 1024) {
                                disruptionsTxt = disruptionsTxt + `[${disruptions[route['route_id']][disruption_type][disruption]['title']}](${disruptions[route['route_id']][disruption_type][disruption]['url']})\n`
                            } else {
                                disruptionEmbed.addField(disruption_type, disruptionsTxt);
                                disruptionsTxt = `[${disruptions[route['route_id']][disruption_type][disruption]['title']}](${disruptions[route['route_id']][disruption_type][disruption]['url']})\n`;
                            }
                        }

                        disruptionEmbed.addField(disruption_type, disruptionsTxt);
                    }
                }

                disruptionsEmbeds[route['route_id']] = disruptionEmbed;
            }
        }

        return disruptionsEmbeds
    }
}

module.exports = ptvFormatter;